### PR TITLE
[AArch64][SVE] Add codegen support for partial reduction lowering to wide add instructions

### PR DESCRIPTION
**Pull request by JamesChesterman**
For partial reductions in the situation of the number of lanes being halved and the bits per lane being doubled, a pair of wide add instructions can be used.

### CODE REVIEW

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (huntergr-arm) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:2045]: I think we may want to start looking at the inputs at this point, as the types that we have instructions for differ based on the chain of operations on the second operand. May be something for a follow-up patch.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (MacDue) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:2047]: I know this was already like this, but I feel like an `is_contained()` check would be more readable than these big expressions :thinking:  
```c++
  EVT VT = EVT::getEVT(I->getType());
  return !is_contained<EVT>({MVT::nxv4i64, MVT::nxv4i32, MVT::nxv2i64,
                             MVT::nxv8i16, MVT::v4i64, MVT::v4i32, MVT::v2i32,
                             MVT::v8i16},
                            VT);
```

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (huntergr-arm) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:21821]: Can remove the `!(a == b)` in favour of `a != b`. `||` is lower precedence so we don't need extra parens here.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (MacDue) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:21806]: nit (optional), I think these names could use some comment shorthands:

- `Accumulator` -> `Acc`
- `AccumulatorType` -> `AccVT`
- `AccumulatorElementType` -> `AccElemVT`
- `InputType` -> `InputVT`

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (huntergr-arm) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:21829]: I think adding dedicated AArch64ISD nodes for these operations would be good, though that can also be in another patch.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (huntergr-arm) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:21797]: Based on the intrinsics you're lowering to below, I think you need to bail if SVE isn't available regardless. There are similar instructions for NEON but they aren't matched here.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (JamesChesterman) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:2045]: Done

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (JamesChesterman) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:21821]: Done

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (JamesChesterman) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:21829]: Will be done in a future PR

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (JamesChesterman) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:21797]: Done

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (JamesChesterman) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp]: Refactored this code so this isn't applicable anymore. I'll bear this function in mind for future development though.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (JamesChesterman) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp]: Done

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (MacDue) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:2048]: The previous check limited this to specific types, this only looks at the element counts, but presumably not all of these can be mapped to instructions. What happens int those cases?

Comment (MacDue) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:2048]: It looks like it should fallback to `DAG.getPartialReduceAdd()` (in `performIntrinsicCombine()`), but `tryLowerPartialReductionToWideAdd()` does not seem to check if there is a wide add instruction available.

Comment (JamesChesterman) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:2048]: I did this with the view of implementing type legalisation I think. For example, we do want a partial reduction from `nxv4i16` to `nxv2i32` (after the input has been extended) to be possible right? And I thought that all that would be needed to be done in checking the wide add instruction is available is checking whether SVE was enabled, or is there something I'm missing?

Comment (JamesChesterman) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:2048]: I think you may be right, it doesn't seem like other intrinsics like dot product have legalisation. So do you think that all the tests required are already in `sve-partial-reduce-wide-add.ll` ? `(`nxv4i32`->nxv2i64`, `nxv8i16`->`nxv4i32`, `nxv16i8`->`nxv8i16`)

Comment (JamesChesterman) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:2048]: This is now resolved with an additional check in the `tryLowerPartialReductionToWideAdd` function.

Comment (MacDue) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:2048]: Cool, thanks!

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (MacDue) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:21831]: nit: I think this check is redundant with the type checks above.

**Review by MacDue — APPROVED** [review]
LGTM (just one nit). I'd wait for @huntergr-arm to approve before landing though :slightly_smiling_face:

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (JamesChesterman) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:21831]: I just thought I'd leave it here ready for the future legalisation work. But I can get rid of it if you like and just put it back in with the future patch?

Comment (MacDue) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:21831]: If the legalization is done before lowering the partial reduction you should have legal types here. I don't think the legalization should be done on the `AArch64::SADDWB` ISD node for example, as I think those would be expected to have 1-to-1 mapping with the instructions.

Comment (JamesChesterman) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp]: Done

**Review by huntergr-arm — APPROVED** [review]
LG

**Pull request merged**
